ethexe/consensus/src/utils.rs

```diff
@@ -0,0 +1,354 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2025 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! # Utilities Module
+//!
+//! This module provides utility functions and data structures for handling batch commitments,
+//! validation requests, and multi-signature operations in the Ethexe system.
+
+use anyhow::Result;
+use ethexe_common::gear::{BatchCommitment, BlockCommitment, CodeCommitment};
+use ethexe_signer::{
+    sha3::digest::Update, Address, ContractSignature, ContractSigner, Digest, PublicKey, ToDigest,
+};
+use gprimitives::H256;
+use parity_scale_codec::{Decode, Encode};
+use std::collections::BTreeMap;
+
+/// Represents a request for validating a batch of block commitments.
+/// This structure is used to verify the integrity and validity of multiple block commitments
+/// and their associated code commitments.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct BatchCommitmentValidationRequest {
+    /// List of block commitment validation requests
+    pub blocks: Vec<BlockCommitmentValidationRequest>,
+    /// List of code commitments to be validated
+    pub codes: Vec<CodeCommitment>,
+}
+
+impl From<&BatchCommitment> for BatchCommitmentValidationRequest {
+    fn from(batch: &BatchCommitment) -> Self {
+        BatchCommitmentValidationRequest {
+            blocks: batch
+                .block_commitments
+                .iter()
+                .map(BlockCommitmentValidationRequest::from)
+                .collect(),
+            codes: batch.code_commitments.clone(),
+        }
+    }
+}
+
+impl ToDigest for BatchCommitmentValidationRequest {
+    fn update_hasher(&self, hasher: &mut ethexe_signer::sha3::Keccak256) {
+        hasher.update(self.codes.to_digest().as_ref());
```

Comment:
why codes go first?

Reply:
Must be compatible with
https://github.com/gear-tech/gear/blob/4a2eaeb5828d91f6f01aea906b69ddb7baba7601/ethexe/signer/src/digest.rs#L208-L219

Reply:
wdyt if we change the order? not a big deal actually, but just for beautification: blocks first, then codes? (as in structure declaration / alphabetical order)

Reply:
and having a test for their compatibility would be great as well

Reply:
Must be compatible with solidity part also. I could add new issue

Reply:
done